### ISSUE TITLE
Idris takes forever/really long to type-check code with where bindings (since 0.9.18)

### ISSUE THREAD

**Issue by timjb** [idris-lang/Idris-dev#2296]
[FingerTree.idr](https://github.com/timjb/idris-pfds/blob/1289557d80e6fbca2f93e9d8b03f79149b647c15/FingerTree.idr) takes forever to type-check in Idris 0.9.18. In version 0.9.17 type-checking that file took about 40 seconds. Incremental commenting in code, it seems like the culprit is this function:

```idris
traverseWPNode : (Measured v1 a1, Measured v2 a2, Applicative f) =>
  (v1 -> a1 -> f a2) -> v1 -> Node v1 a1 -> f (Node v2 a2)
traverseWPNode f v (Node2 _ a b) = [| node2 (f v a) (f va b) |]
  where va = v <+> measure a
traverseWPNode f v (Node3 _ a b c) = [| node3 (f v a) (f va b) (f vab c) |]
  where va  = v <+> measure a
        vab = va <+> measure b
```

When commenting out this function and the rest of the file, Idris 0.9.18 takes 2 seconds to type-check. With this code enabled, Idris takes forever (more than 5 minutes). Curiously, the following version of `traverseWPNode` with the where-clauses inlined type-checks in two-seconds:

```idris
traverseWPNode : (Measured v1 a1, Measured v2 a2, Applicative f) =>
  (v1 -> a1 -> f a2) -> v1 -> Node v1 a1 -> f (Node v2 a2)
traverseWPNode f v (Node2 _ a b) = [| node2 (f v a) (f (v <+> measure a) b) |]
traverseWPNode f v (Node3 _ a b c) = [| node3 (f v a) (f (v <+> measure a) b) (f (v <+> measure a <+> measure b) c) |]
```

**Comment by edwinb** [comment]
Without having looked - are there lots of ambiguous names here? That's
usually the reason for slow type checking on functions of this size.
Using 'with' clauses to disambiguate may speed things up...

I wouldn't be surprised if that was why it was much slower since there
were ambiguity problems that have been resolved in 0.9.18 but perhaps it
is at the cost of retrying too many things.

In any case, I'll take a closer look when I get the chance.

**Comment by timjb** [comment]
I don't see any ambiguous names in `traverseWPNode`.

Every function called in `traverseWPNode` ( `node2`, `node3`, `<+>`, `<$>` and `<*>`) has been used before in another function. There's nothing suspicious about `traverseWPNode` from this perspective.

**Comment by edwinb** [comment]
Well I just tried it with logging and there is nothing immediately obvious, and it is certainly the elaboration which is slow rather than any of the obvious things I've seen before, so this will be an interesting one to poke further. 

I don't know if I will have much time to do this in the next few days I'm afraid, but this sort of slowness bothers me so I won't leave this for too long.

Some workarounds for now may be to write a version of <$> with specific types rather than using the type class variant (and possibly the other things).

By the way, a quick way to find out where the bottleneck is would be to run Idris with --log 1, then look for where the long pause is... logging won't necessarily help find the solution but at least you can quickly find out which definition is the slow one.

**Comment by edwinb** [comment]
Okay, with a bit of poking I've found the cause of the worse performance, if not the cause of the bad performance in the first place. I increased the type class resolution search depth, which has not affected anything else, but this example seems to have some interesting behaviour which requires it to search the entire space before finding the answer. 

Clearly I'll need to probe a bit deeper. I will do this soon but can make no promises as to when (except that I'm sure this will annoy me enough that it won't take too long...)
